sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -218,6 +280,49 @@ object Hash {
             acc.addTypedValue(x._1).addTypedValue(x._2)
           )
       }
+
+    private def addTypedValue(fieldNumber: Int, value: Value): this.type =
+      value match {
+        case Value.ValueUnit =>
+          this
+        case Value.ValueBool(b) =>
+          add(fieldNumber, b)
+        case Value.ValueInt64(v) =>
+          add(fieldNumber, v)
+        case Value.ValueNumeric(v) =>
+          add(fieldNumber, v)
+        case Value.ValueTimestamp(v) =>
+          add(fieldNumber, v.micros)
+        case Value.ValueDate(v) =>
+          add(fieldNumber, v.days)
+        case Value.ValueParty(v) =>
+          // No default
+          add(fieldNumber).add(v)
+        case Value.ValueText(v) =>
+          add(fieldNumber, v)
+        case Value.ValueContractId(cid) =>
+          // No default
+          add(fieldNumber).addCid(cid)
+        case Value.ValueOptional(opt) =>
+          iterateOver(fieldNumber, opt)(_ addTypedValue _)
+        case Value.ValueList(xs) =>
+          iterateOver(fieldNumber, xs.toImmArray)(_ addTypedValue _)
+        case Value.ValueTextMap(xs) =>
+          iterateOver(fieldNumber, xs.toImmArray)((acc, x) => acc.add(x._1).addTypedValue(x._2))
+        case Value.ValueRecord(_, fs) =>
+          // No default
+          add(fieldNumber).addRecord(fs)
```

Comment:
This looks odd. Why did you choose to not add a default for records?

AFAICS we only need the default for `Optional` to support the current upgrading features. So all the other defaults (strings, numbers, ...) are some for of future-proofing. But then, why not consider the record whose fields only contain default values as the default?

Reply:
We have to take into account that every change we would make to the hashing scheme would be breaking. So I think it is worthwhile to handle default value for more values than just `Optional`, so it give us more flexibility in the future if we want to extends the default value.

To do so I take inspiration from [protobuf default values](https://protobuf.dev/programming-guides/proto3/#default):
>For strings, the default value is the empty string.
    For bytes, the default value is empty bytes.
    For bools, the default value is false.
    For numeric types, the default value is zero.
    For enums, the default value is the first defined enum value, which must be 0.
    For message fields, the field is not set. Its exact value is language-dependent. See the [generated code guide](https://protobuf.dev/reference/) for details.

In particular there is not default value for
- equivalent of record: an message without field (what would be the closer to a empty record).
- equivalent of variant: or a message containing only one "one of" where the field number '0' is set to a default value.
Both are encoded as an empty message prefixed by its length (i.e. the bytes [0x00],  the [varint](https://protobuf.dev/programming-guides/encoding/#varints) encoding of `0` follow by `[]`)  

Since we do not have the position of the enum constructor I cannot easily check for its default value. 
(Maybe I will have time to handle that latter).

proto equivalent of `optional`, `list` and `map` have default value as they are message with a unique optional or repeated field.

This answer both question you add.
why default for something else than `Optional` and why no default for record.

Reply:
Protobuf encodings are a good starting point. Extensible records in Daml indeed correspond to Protobuf. And if you look at the [encoding rules for messages](https://protobuf.dev/programming-guides/encoding/#cheat-sheet), you'll see that messages where all fields contain the default values (and none of them are marked as `optional` or `repeated`) indeed yield the empty bytestring. And this empty bytestring is then again considered as the default value in larger message structures. So in this sense, Protobuf does have default values for messages and one could imagine a default value for records as well.

Overall, I think it makes sense to define default values for non-`optional` values too. We just need to be careful to not shoot into our foots w.r.t. hash collisions (or equivalently ambiguous encodings).